vllm/distributed/device_communicators/all2all.py

```diff
@@ -91,3 +125,86 @@ def combine(self, hidden_states: torch.Tensor) -> torch.Tensor:
         pass
+
+
+class PPLXAll2All(All2AllBase):
+    """
+    All2All communication based on PPLX kernels.
+    """
+
+    def __init__(self, cpu_group, model: torch.nn.Module):
+        has_pplx = importlib.util.find_spec("pplx_kernels") is not None
+        assert has_pplx, "pplx_kernels not found. Please follow https://github.com/vllm-project/vllm/blob/main/tools/ep_kernels/README.md to install pplx_kernels."  # noqa
+        import pplx_kernels as pplx
+        super().__init__(cpu_group, model)
+
+        if self.internode:
+            # inter-node communication needs nvshmem,
+            # intra-node communication uses p2p mapping directly
+            from pplx_kernels.nvshmem import (nvshmem_alloc_empty_unique_id,
+                                              nvshmem_get_unique_id,
+                                              nvshmem_init)
+            logger.debug(
+                "Initialize NVSHMEM for pplx_kernels: "
+                "rank=%d, world size=%d", self.rank, self.world_size)
+            uid = nvshmem_get_unique_id(
+            ) if self.rank == 0 else nvshmem_alloc_empty_unique_id()
+            dist.broadcast(uid,
+                           src=self.ep_group.ranks[0],
+                           group=self.cpu_group)
+            logger.debug("PPLX NVSHMEM UID = %s", uid)
+            nvshmem_init(uid, self.rank, self.world_size)
+
+        self.handle_cache = Cache()
+
+        moe_modules = [
+            module for module in model.modules()
+            if module.__class__.__name__ == "FusedMoE"
+        ]
+        for module in moe_modules:
+            moe_layer = module
+            moe = moe_layer.moe_config
+            max_num_tokens = moe.max_num_tokens
+
+            all_to_all_args = dict(
+                max_num_tokens=max_num_tokens,
+                num_experts=moe.num_experts,
+                experts_per_token=moe.experts_per_token,  # topk
+                rank=self.rank,
+                world_size=self.world_size,
+                # dp_size actually means tp_size, bug in pplx kernels
+                dp_size=self.tp_group.world_size,
+                hidden_dim=moe.hidden_dim,
+                hidden_dim_bytes=moe.hidden_dim * moe.in_dtype.itemsize,
+                # For blocked per token: set to
+                #   ceil_div(hidden_dim, block_size) * sizeof(float32)
+                # For per-token: set to sizeof(float32)
+                hidden_dim_scale_bytes=(0 if moe.in_dtype.itemsize != 1 else (
+                    (moe.hidden_dim + moe.block_size - 1) // moe.block_size *
+                    torch.float32.itemsize)),
+                group_name=self.cpu_group.group_name,
+            )
+
+            pplx_handle = self.handle_cache.get_or_create(
+                all_to_all_args, pplx.AllToAll.internode
+                if self.internode else pplx.AllToAll.intranode)
+
+            moe_layer.quant_method.fused_experts.prepare_finalize.a2a \
+                = pplx_handle
```

Comment:
I think it would be cleaner to call `_construct_prepare_finalize` and stash the result in `moe_layer.quant_method.fused_experts.prepare_finalize` rather than poking the a2a into the already constructed object.

Maybe make the whole process into a method, e.g. move the `_construct_prepare_finalize` call + `set_prepare_finalize` into a single `init_prepare_finalize` method.

Reply:
I also am generally opposed to shoving things into objects like this, but I do wonder sometimes if I'm colored by my C++ background and should embrace Python's "flexibility" more

@bnellnm what would that look like? Maybe we try it in a followup?

Reply:
merged them into `init_prepare_finalize` now.